### ISSUE TITLE
Boundary of an ill-typed equality?

### ISSUE THREAD

**Issue by mikeshulman** [Andromedans/andromeda#488]
The following:

```
rule A type
rule B (x:A) type
rule C (x:A) (y:B x) type
rule a1 : A
rule a2 : A
rule ae : a1 == a2 : A
rule ae' : a2 == a1 : A
rule b1 : B a1
rule b2 : B a2
rule be : b1 == convert b2 (congruence (B a2) (B a1) ae') : B a1

rule Dummy type
rule dummy : Dummy
rule dummy_eq : dummy == dummy : Dummy ;;

handle
  congruence (C a1 b1) (C a2 b2) ae dummy_eq
with
| ML.coerce _ (z1 == z2 : Z as ??) ->
  base.print (z2 : Z);
  yield (ML.Some be)
end
```

produces `Runtime error: unhandled operation ML.coerce (⊢ b2) (⊢ ⁇ : B a1)` (from the `print` line -- without that line it works fine).  This confuses me; I expected that in a boundary of a term equality I could trust that the two terms involved both belong to the specified type.  Andromeda doesn't seem to let _me_ create such a boundary in which the terms don't both belong to the specified type; if I write `b1 == b2 : B a1 as ??` I get the same error `Runtime error: unhandled operation ML.coerce (⊢ b2) (⊢ ⁇ : B a1)`.  But it seems that it can produce such boundaries internally and give them to me?  Is that the intended behavior?

**Comment by andrejbauer** [comment]
You're correct. This is flawed design, it's also a fairly easy fix. I am classifying it as a bug.

**Comment by andrejbauer** [comment]
Closed by e4f09f02078666a62de1c206d95fcf0ff1388eeb
